core/commands/repo.go

```diff
@@ -0,0 +1,104 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	cmds "github.com/jbenet/go-ipfs/commands"
+	"github.com/jbenet/go-ipfs/core"
+	u "github.com/jbenet/go-ipfs/util"
+)
+
+var RepoCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Manipulate the IPFS repo",
+		ShortDescription: `
+'ipfs repo' is a plumbing command used to manipulate the repo.
+`,
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"gc": repoGcCmd,
+	},
+}
+
+type KeyRemoved struct {
+	Key u.Key
+}
+
+var repoGcCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Perform a garbage collection sweep on the repo",
+		ShortDescription: `
+'ipfs repo gc' is a plumbing command that will sweep the local
+set of stored objects and remove ones that are not pinned in 
+order to reclaim hard disk space.
+`,
+	},
+
+	Options: []cmds.Option{
+		cmds.BoolOption("quiet", "q", "Write minimal output"),
+	},
+	Run: func(req cmds.Request) (interface{}, error) {
+		n, err := req.Context().GetNode()
+		if err != nil {
+			return nil, err
+		}
+
+		keychan, err := n.Blockstore.AllKeysChan(req.Context().Context, 0, 1<<16)
+		if err != nil {
+			return nil, err
+		}
+
+		outChan := make(chan interface{})
+		go GarbageCollectBlockstore(n, keychan, outChan)
+		return outChan, nil
+	},
+	Type: KeyRemoved{},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			outChan, ok := res.Output().(chan interface{})
+			if !ok {
+				return nil, u.ErrCast()
+			}
+
+			quiet, _, err := res.Request().Option("quiet").Bool()
+			if err != nil {
+				return nil, err
+			}
+
+			marshal := func(v interface{}) (io.Reader, error) {
+				obj, ok := v.(*KeyRemoved)
+				if !ok {
+					return nil, u.ErrCast()
+				}
+
+				var buf *bytes.Buffer
+				if quiet {
+					buf = bytes.NewBufferString(string(obj.Key) + "\n")
+				} else {
+					buf = bytes.NewBufferString(fmt.Sprintf("removed %s\n", obj.Key))
+				}
+				return buf, nil
+			}
+
+			return &cmds.ChannelMarshaler{
+				Channel:   outChan,
+				Marshaler: marshal,
+			}, nil
+		},
+	},
+}
+
+func GarbageCollectBlockstore(n *core.IpfsNode, keychan <-chan u.Key, output chan interface{}) {
```

Comment:
For the sake of code reuse, perhaps this free function should be placed in a package under `core` so it can be used in other contexts.

Recently, we added a `coreio` package which houses `Add` and `Cat` operations. 

https://github.com/jbenet/go-ipfs/tree/master/core/io

Maybe this would be do well in a `coredata` (or something or the other). It might look something like this...

```
package data

import (
	"code.google.com/p/go.net/context"
	"github.com/jbenet/go-ipfs/core"
	"github.com/jbenet/go-ipfs/util"
)

type GarbageCollectResults struct {
	removed chan util.Key
	Err     error
}

// Removed returns keys as they are removed. Channel is closed when operation
// is complete or in the event of an error. In the event of an error,
// GarbageCollectResults.Err is non-nil.
func (r *GarbageCollectResults) Removed() <-chan util.Key {
	return r.removed
}

func GarbageCollectAsync(ctx context.Context, n *core.IpfsNode) *GarbageCollectResults {
	// always return r.
	r := &GarbageCollectResults{
		removed: make(chan util.Key),
		Err:     nil, // initially
	}
	defer close(r.removed) // always
	allKeys, err := n.Blockstore.AllKeysChan(ctx, 0, 1<<16)
	if err != nil {
		r.Err = err
		return r
	}
	for {
		select {
		case k := <-allKeys:
			if !n.Pinning.IsPinned(k) {
				err := n.Blockstore.DeleteBlock(k)
				if err != nil {
					r.Err = ctx.Err()
					return r // Blockstore errors are permanent/non-recoverable
				}
				select {
				case r.removed <- k:
				case <-ctx.Done():
					r.Err = ctx.Err()
					return r
				}
			}
		case <-ctx.Done():
			r.Err = ctx.Err()
			return r
		}
	}
	return r
}
```